exercises/linked-list/linked_list_test.go

```diff
@@ -0,0 +1,135 @@
+package linkedlist
+
+import (
+	"bytes"
+	"fmt"
+	"testing"
+)
+
+func TestNodeTraversal(t *testing.T) {
+	ll := NewList(1, 2, 3, 4, 5)
+	node := ll.head.next.next
+
+	if want, got := ll.head, node.First(); want != got {
+		t.Errorf("%#v.First() is wrong, expected= %#v, got= %#v", node, want, got)
+	}
+	if want, got := ll.tail, node.Last(); want != got {
+		t.Errorf("%#v.Last() is wrong, expected= %#v, got= %#v", node, want, got)
+	}
+	if want, got := ll.tail.prev, node.Next(); want != got {
+		t.Errorf("%#v.Next() is wrong, expected= %#v, got= %#v", node, want, got)
+	}
+	if want, got := ll.head.next, node.Prev(); want != got {
+		t.Errorf("%#v.Prev() is wrong, expected= %#v, got= %#v", node, want, got)
+	}
+}
+
+func TestListTraversal(t *testing.T) {
+	ll := NewList(1, 2, 3, 4, 5)
+
+	if want, got := ll.head, ll.First(); want != got {
+		t.Errorf("%#v.First() is wrong, expected= %#v, got= %#v", ll, want, got)
+	}
+	if want, got := ll.tail, ll.Last(); want != got {
+		t.Errorf("%#v.First() is wrong, expected= %#v, got= %#v", ll, want, got)
+	}
+}
+
+func TestNew(t *testing.T) {
+	for _, tc := range newListTestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := NewList(tc.in...)
+
+			checkDoublyLinkedList(t, actual, tc.expected)
+		})
+	}
+}
+
+func TestReverse(t *testing.T) {
+	for _, tc := range reverseTestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := NewList(tc.in...)
+			actual.Reverse()
+
+			checkDoublyLinkedList(t, actual, tc.expected)
+		})
+	}
+}
+
+func TestPushPop(t *testing.T) {
+	for _, tc := range pushPopTestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := NewList(tc.in...)
+			for _, ac := range tc.actions {
+				ac(t, actual)
+			}
+
+			checkDoublyLinkedList(t, actual, tc.expected)
+		})
+	}
+}
+
+// checkDoublyLinkedList checks that the linked list is constructed correctly.
+func checkDoublyLinkedList(t *testing.T, ll *List, expected []interface{}) {
+	// check that length and elements are correct (scan once from begin -> end)
+	elem, count, idx := ll.head, 0, 0
```

Comment:
Perhaps
```
-    elem, count, idx := ll.head, 0, 0
+    elem, count, idx := ll.First(), 0, 0
```
This gives some lattitude to the implementor instead of requiring an exact duplication of your example struct(s).